### PR TITLE
Prettier printing

### PR DESCRIPTION
**Pull request by bakkot**
Pretty-printing mode now has indentation and spaces, making it reasonably close to the code humans would write in most cases, addressing #26. Also adds a `forEach` method on CodeRep elements for easier manipulation (e.g. this is what makes indentation possible).

Note that this PR adds CodeRep elements to the minimal CodeGen which do not cause any additional output unless overridden. This seemed a better solution than overriding every method with an almost identical one containing formatting information.

Output looks like this:

``` js
class FormattedCodeGen extends CodeGen {
  commaSep(pieces) {
    let first = true;
    pieces = pieces.map(p => {
      if (first) {
        first = false;
        return p;
      } else {
        return seq(this.space(), p);
      }
    });
    return new CommaSep(pieces);
  }
}
```

### CODE REVIEW

**Comment by michaelficarra** [comment]
:worried: I really don't like this. I don't want the minimal code generator to deal with stylistic whitespace at all. Additionally, this isn't really flexible; I couldn't define my own preferred formatting without overriding each method.

Even though it's a bit verbose and there will surely be plenty of duplication, I'd rather see each method overridden in the pretty formatter.

**Comment by bakkot** [comment]
I can do that, but it's going to approximately double the size of the code. An outright majority of methods will be overwritten by methods which are >95% identical. Any changes to either generator will need to be duplicated in the other.

As it stands, the minimal codegen already inserts bunches of `Empty` CodeReps. What's different about this? It's just more empty CodeReps, unless you subclass it. I strongly prefer code designed to be subclassed.

The alternative I was considering was to defer to `super` methods and then insert spaces and linebreaks in the resulting `Seq` CodeReps, but that's going to be extremely fragile and difficult to understand.

**Comment by bakkot** [comment]
To improve extensibility, I can change all of the `this.space()`s and `this.linebreak()`s to `this.separator(AFTER_IF)` etc, where the default behavior of `separator` is to return `Empty` for all inputs. Subclasses doing formatting can overwrite just that method and modify its return value for a given input as appropriate.

**Comment by bakkot** [comment]
Per discussion, there's now a MinimalCodeGen and an ExtensibleCodeGen, which duplicate most of their code. FormattedCodeGen subclasses ExtensibleCodeGen.

File: src/formatted-codegen.js [hunk 1]
```diff
@@ -0,0 +1,1100 @@
...
+
+  brace(rep, node) {
+    return new Brace(rep);
+  }
+
+  bracket(rep) {
+    return new Bracket(rep);
+  }
+
+  commaSep(pieces, separator) {
```
Comment (michaelficarra) [inline src/formatted-codegen.js]: This needs separators for both before and after the commas.

File: src/formatted-codegen.js [hunk 2]
```diff
@@ -0,0 +1,1100 @@
...
+  reduceTemplateExpression(node, {tag, elements}) {
+    let state = node.tag == null ? empty() : this.p(node.tag, getPrecedence(node), tag);
+    let templateData = "";
+    state = seq(state, this.t("`"));
+    for (let i = 0, l = node.elements.length; i < l; ++i) {
+      if (node.elements[i].type === "TemplateElement") {
+        let d = "";
+        if (i > 0) d += "}";
+        d += node.elements[i].rawValue;
+        if (i < l - 1) d += "${"
```
Comment (michaelficarra) [inline src/formatted-codegen.js:928]: We need before and after interpolated expressions.

File: src/formatted-codegen.js [hunk 3]
```diff
@@ -0,0 +1,1100 @@
...
+};
+
+Sep.BEFORE_POSTFIX = function(op) {
+  return {
+    type: "BEFORE_POSTFIX",
+    op: op
+  };
+};
+
+Sep.UNARY = function(op) {
```
Comment (michaelficarra) [inline src/formatted-codegen.js:305]: How do users distinguish unary prefix and unary postfix?

File: src/formatted-codegen.js [hunk 4]
```diff
@@ -0,0 +1,1100 @@
...
+    if (!node.isStatic) return method;
+    return seq(this.t("static"), this.sep(Sep.AFTER_STATIC), method);
+  }
+
+  reduceComputedMemberExpression(node, {object, expression}) {
+    let startsWithLetSquareBracket =
+      object.startsWithLetSquareBracket ||
+      node.object.type === "IdentifierExpression" && node.object.name === "let";
+    return objectAssign(
+      seq(this.p(node.object, getPrecedence(node), object), this.bracket(expression)),
```
Comment (michaelficarra) [inline src/formatted-codegen.js]: Missing space between object and bracket.

Comment (michaelficarra) [inline src/formatted-codegen.js]: Now that I think about it, spaces inside the brackets are also needed.

File: src/formatted-codegen.js [hunk 5]
```diff
@@ -0,0 +1,1100 @@
...
+    return a;
+  }
+
+  reduceIfStatement(node, {test, consequent, alternate}) {
+    if (alternate && consequent.endsWithMissingElse) {
+      consequent = this.brace(consequent, node);
+    }
+    return objectAssign(
+      seq(this.t("if"), this.sep(Sep.AFTER_IF),
+        this.paren(test), this.sep(Sep.AFTER_IF_TEST),
```
Comment (michaelficarra) [inline src/formatted-codegen.js]: Needs spaces inside these parentheses.

File: src/formatted-codegen.js [hunk 6]
```diff
@@ -0,0 +1,1100 @@
...
+    }
+    return original;
+  }
+
+  t(token) {
+    return new Token(token);
+  }
+
+  p(node, precedence, a) {
+    return getPrecedence(node) < precedence ? this.paren(a) : a;
```
Comment (michaelficarra) [inline src/formatted-codegen.js]: Needs spaces on either side of parentheses used for grouping.

**Comment by michaelficarra** [comment]
I didn't want to comment on them all individually, but there's a lot of missing positions. You should try to mark every position where whitespace could be used, regardless of whether it is common, so that the extensible codegen is as useful as possible.

I think a good test would be something like everything.js except spaces _everywhere possible_.

**Comment by bakkot** [comment]
I'll see what I can do, although I would say that cluttering it with markers no one will use will make the extensible codegen less useful, not more.

Also, some places there will be a lot of redundancy: for example, there can be a space before and after each comma in an array literal and also both before and after each element (which are not equivalent: not every element is preceded or followed by a comma) and also before and after both braces. Should these all have distinct markers?

I suppose in that case it suffices to have a separators before and after the brackets and before and after the commas, and a separate one for empty elements which will mostly not be redundant. There's probably similarly complex cases, though.

Also, strictly speaking, there can be a space before any statement as well as after, and also before and after any of `[...]`, `(...)`, `{...}`. Do we want to have markers for all of those, and for other such it's-technically-legal cases?

File: src/formatted-codegen.js (hunk 3 above)
Comment (bakkot) [inline src/formatted-codegen.js:305]: Unary prefix is grouped with the other unary operators. Unary postfix has its own marker.

Comment (michaelficarra) [inline src/formatted-codegen.js:305]: I see now. Thanks.

**Comment by bakkot** [comment]
Proposed revision: brace, bracket, and parens get distinct separators, per use, for initial, final, and empty separator positions. Commas get separators before as well as after. Statement-terminating semicolons get separators before as well as after. I'll add the couple of spots I overlooked - in property access and templates, etc.

Nodes do NOT insert separators before and after the coderep they generate, with the exception of AFTER_STATEMENT. I believe leaving them out won't make it impossible to add spaces anywhere, with the exception that it will be impossible to begin with whitespace (without subclassing), although it will make it harder to add spaces in some places and not others. I could also include these, at the price of even more bloat: a BEFORE_STATEMENT, separators before and after brackets as well as inside, etc. I would strongly prefer they be left out from a style perspective, but it wouldn't be that much effort to put them in.

**Comment by michaelficarra** [comment]
> Also, strictly speaking, there can be a space before any statement as well as after

I think these positions between statements are the only exceptions I'd be okay with.

> and also before and after any of `[...]`, `(...)`, `{...}`

Those are characters, not positions in the grammar. Not all parentheses or all braces or all brackets are the same. That's why I recommended distinguishing parentheses used for grouping in the `p` function above.

**Comment by bakkot** [comment]
Sorry, I wasn't clear: each distinct use of brackets (etc) would have a distinct initial and final separator. This would probably be implemented by having `this.bracket(rep)` calls be changed to `this.bracket(rep, ARRAY_INITIAL, ARRAY_FINAL, ARRAY_EMPTY)` and so on.

**Comment by michaelficarra** [comment]
Sounds good, but lose the statement stuff.

File: package.json [hunk 7]
```diff
@@ -15,9 +15,9 @@
   "scripts": {
     "test": "mocha --inline-diffs --check-leaks --ui tdd --reporter dot --timeout 5000 --recursive test",
     "build": "babel --auxiliary-comment-before 'istanbul ignore next' --modules commonStrict --source-maps-inline --out-dir dist src",
-    "generate-coverage": "istanbul cover _mocha -- --inline-diffs --check-leaks --ui tdd --reporter dot --timeout 5000 --recursive test",
+    "generate-coverage": "istanbul cover -x '**/dist/index.js' _mocha -- --inline-diffs --check-leaks --ui tdd --reporter dot --timeout 5000 --recursive test",
```
Comment (michaelficarra) [inline package.json]: Why is this `-x` needed?

Comment (bakkot) [inline package.json]: The file Babel generates is only covered if every exported name `require`d somewhere. Some names are made available so new generators can be written, but not imported by the test suite. See src/index.is.

Comment (michaelficarra) [inline package.json]: I'll take your word for it, but can't we just use `dist/index.js`?

Comment (bakkot) [inline package.json]: Apparently yes! Corrected.

**Pull request merged**